### PR TITLE
Added disk utilities to create and delete partition on disk

### PR DESCRIPTION
**Pull request by maramsmurthy**
create_linux_raw_partition: Creates specified number of partitions
get_size_of_disk: Retruns size of disk in bytes
delete_partition: Deletes specified partition
clean_disk: Deletes partitions and wipes partition table
get_disk_partitions: Returns exisitng partition on disk

### CODE REVIEW

**Comment by maramsmurthy** [comment]
[root@ltcrain7zx-lp3 disk]# avocado run --max-parallel-tasks=1 usb_tests.py -m usb_tests.py.data/usb_tests.yaml 
JOB ID     : 45f5d207356c3c651a1ce987768838a2d09be2fd
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-06T12.14-45f5d20/job.log
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-378f: STARTED
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-378f: PASS (5.89 s)
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-378f: STARTED
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-378f: PASS (5.80 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-06T12.14-45f5d20/results.html
JOB TIME   : 30.53 s

**Comment by maramsmurthy** [comment]
Ran test on USB disk
[root@ltcden4-lp7-murthy disk]# avocado run --max-parallel-tasks=1 usb_tests.py:test_create_usb_partitions -m usb_tests.py.data/usb_tests.yaml
JOB ID     : 7251346db1e66d00573132d44492715e0063cb0c
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-07T15.04-7251346/job.log
 (1/1) usb_tests.py:USBTests.test_create_usb_partitions;run-ae01: STARTED
 (1/1) usb_tests.py:USBTests.test_create_usb_partitions;run-ae01: PASS (7.35 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-07T15.04-7251346/results.html
JOB TIME   : 9.07 s
[root@ltcden4-lp7-murthy disk]# sfdisk -l /dev/sdf
Disk /dev/sdf: 149 GiB, 160037691392 bytes, 312573616 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0xe1f871c8

Device     Boot     Start       End   Sectors  Size Id Type
/dev/sdf1            2048  20973567  20971520   10G 83 Linux
/dev/sdf2        20973568  41945087  20971520   10G 83 Linux
/dev/sdf3        41945088  62916607  20971520   10G 83 Linux
/dev/sdf4        62916608 312573615 249657008  119G  5 Extended
/dev/sdf5        62918656  83890175  20971520   10G 83 Linux
/dev/sdf6        83892224 104863743  20971520   10G 83 Linux
/dev/sdf7       104865792 125837311  20971520   10G 83 Linux
/dev/sdf8       125839360 146810879  20971520   10G 83 Linux
/dev/sdf9       146812928 167784447  20971520   10G 83 Linux

[root@ltcden4-lp7-murthy disk]# avocado run --max-parallel-tasks=1 usb_tests.py:test_delete_all_usb_partitions -m usb_tests.py.data/usb_tests.yaml
JOB ID     : 017655e7b5b64f9b2911504d464e879d013ef02c
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-07T15.05-017655e/job.log
 (1/1) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-ae01: STARTED
 (1/1) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-ae01: PASS (5.87 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-07T15.05-017655e/results.html
JOB TIME   : 7.42 s
[root@ltcden4-lp7-murthy disk]# sfdisk -l /dev/sdf
Disk /dev/sdf: 149 GiB, 160037691392 bytes, 312573616 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
[root@ltcden4-lp7-murthy disk]#

**Comment by maramsmurthy** [comment]
On USB Disk
disk: /dev/sdf
[root@ltcden4-lp7-murthy disk]# avocado run --max-parallel-tasks=1 usb_tests.py -m usb_tests.py.data/u
sb_tests.yaml
JOB ID     : bc124d9eef781da3ab31570e462aabcd2f5251e0
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-08T17.08-bc124d9/job.log
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-845d: STARTED
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-845d: PASS (2.01 s)
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-845d: STARTED
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-845d: PASS (1.27 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-08T17.08-bc124d9/results.html
JOB TIME   : 6.76 s

disk: /dev/disk/by-id/usb-IBM_T_RDX-USB3_00102000195F-0:0
[root@ltcden4-lp7-murthy disk]# avocado run --max-parallel-tasks=1 usb_tests.py -m usb_tests.py.data/usb_tests.yaml                                                                                        
JOB ID     : 066a6fa759ad168678a1aedcae9598694ce04d52
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-08T17.16-066a6fa/job.log
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-b09d: STARTED
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-b09d: PASS (2.82 s)
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-b09d: STARTED
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-b09d: PASS (1.99 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-08T17.16-066a6fa/results.html
JOB TIME   : 7.97 s

disk: /dev/disk/by-path/pci-0154:80:00.0-usb-0:4:1.0-scsi-0:0:0:0
[root@ltcden4-lp7-murthy disk]# avocado run --max-parallel-tasks=1 usb_tests.py -m usb_tests.py.data/usb_tests.yaml                                                                                        
JOB ID     : 2b509c2554ba305c12fe0484858c2053f05fd6b7
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-08T17.23-2b509c2/job.log
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-e508: STARTED
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-e508: PASS (3.23 s)
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-e508: STARTED
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-e508: PASS (1.85 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-08T17.23-2b509c2/results.html
JOB TIME   : 7.83 s

File: avocado/utils/disk.py [hunk 1]
```diff
@@ -311,3 +312,101 @@ def get_disk_mountpoint(device):
...
+    :type size: str
+    :param num_of_par: Number of partitions to be created
+    :type num_of_par: int
+
+    Returns list of created partitions
+    """
+    if not size:
+        size = get_size_of_disk(disk_name) / 1024 / 1024 / 1024
+        size = size / num_of_par
+        size = str(size) + "G"
```
Comment (abdhaleegit) [inline avocado/utils/disk.py:333]: so if my disk size is 1TB,  size  = 1TB/(1024 * 3) --> instead of /1024/1024/1024  ? just think if feasible ?

Comment (abdhaleegit) [reply avocado/utils/disk.py:333]: @maramsmurthy a check for smaller disks also needed.. I mean what if usb is 20G

File: avocado/utils/disk.py [hunk 2]
```diff
@@ -311,3 +312,101 @@ def get_disk_mountpoint(device):
...
+    ]
+    disk_partition_file = (
+        "/tmp/creat_partition" + process.run("date '+%d-%m-%y_%T'").stdout_text.strip()
+    )
+    if not os.path.isfile(disk_partition_file):
+        process.run("touch " + disk_partition_file)
+    for line in partitions:
+        genio.append_one_line(disk_partition_file, line)
+    part_output = process.getoutput("sfdisk " + disk_name + " < " + disk_partition_file)
+    for _ in range(60):
```
Comment (abdhaleegit) [inline avocado/utils/disk.py:345]: range is hardcoded to 60 .. is that max partitions ?

**Review by abdhaleegit — CHANGES_REQUESTED** [review]

File: avocado/utils/disk.py (hunk 2 above)
Comment (maramsmurthy) [reply avocado/utils/disk.py:345]: Hi Abdul! Used this to avoid sleep time.

File: avocado/utils/disk.py (hunk 1 above)
Comment (maramsmurthy) [reply avocado/utils/disk.py:333]: "get_size_of_disk" definition returns size in number of bytes. We are just converting it to GB's. Not performing any operation here

File: avocado/utils/disk.py (hunk 2 above)
Comment (maramsmurthy) [reply avocado/utils/disk.py:345]: No Abdul! It's not max partitions. Just took a random number to wait for some time for the condition to be true

Comment (abdhaleegit) [reply avocado/utils/disk.py:345]:
```
flag = True
while flag:
    if <check something>:
         flag = False
         
```

File: avocado/utils/disk.py [hunk 3]
```diff
@@ -311,3 +312,101 @@ def get_disk_mountpoint(device):
...
+    :type disk_name: str
+    """
+    output = process.getoutput("sfdisk --delete " + disk_name)
+    for _ in range(60):
+        if int(process.getoutput("ls " + disk_name + "* | wc -l")) == 1:
+            break
+    while not get_disk_partitions(disk_name):
+        if "The partition table has been altered" in output:
+            process.run("wipefs -af " + disk_name)
+            break
```
Comment (abdhaleegit) [inline avocado/utils/disk.py:394]: check if above while is needed ?

**Comment by maramsmurthy** [comment]
disk: "/dev/disk/by-path/pci-0154:80:00.0-usb-0:4:1.0-scsi-0:0:0:0"
[root@ltcden4-lp7-murthy disk]# avocado run --max-parallel-tasks=1 usb_tests.py -m usb_tests.py.data/usb_tests.yaml 
JOB ID     : 9e59c82d2078f5c7a0862724b8313b2e3cecfb22
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-27T15.03-9e59c82/job.log
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-97b6: STARTED
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-97b6: PASS (17.39 s)
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-97b6: STARTED
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-97b6: PASS (1.31 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-27T15.03-9e59c82/results.html
JOB TIME   : 135.23 s
[root@ltcden4-lp7-murthy disk]# avocado run --max-parallel-tasks=1 usb_tests.py:test_create_usb_partitions -m usb_tests.py.data/usb_tests.yaml                                                              
JOB ID     : e7c5fba26989f30341b492c8938cd2917764f7ff
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-27T15.06-e7c5fba/job.log
 (1/1) usb_tests.py:USBTests.test_create_usb_partitions;run-97b6: STARTED
 (1/1) usb_tests.py:USBTests.test_create_usb_partitions;run-97b6: PASS (2.48 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-27T15.06-e7c5fba/results.html
JOB TIME   : 17.85 s
[root@ltcden4-lp7-murthy disk]# avocado run --max-parallel-tasks=1 usb_tests.py:test_delete_all_usb_partitions -m usb_tests.py.data/usb_tests.yaml                                                          
JOB ID     : 4f82fd01980c277905a39e0cbe0630fdae0aba83
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-27T15.07-4f82fd0/job.log
 (1/1) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-97b6: STARTED
 (1/1) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-97b6: PASS (1.27 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-27T15.07-4f82fd0/results.html
JOB TIME   : 13.52 s

[root@ltcden4-lp7-murthy disk]# cat usb_tests.py.data/usb_tests.yaml 
disk: "/dev/mapper/mpathc"
pci_device: "0131:50:00.0"
num_of_partitions: 8
partition_size: "1G"
[root@ltcden4-lp7-murthy disk]# avocado run --max-parallel-tasks=1 usb_tests.py -m usb_tests.py.data/usb_tests.yaml
JOB ID     : 6660df3df2ff3c38d2a8bd75438dce70b41a4b78
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2023-09-27T22.52-6660df3/job.log
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-3f2b: STARTED
 (1/2) usb_tests.py:USBTests.test_create_usb_partitions;run-3f2b: PASS (0.94 s)
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-3f2b: STARTED
 (2/2) usb_tests.py:USBTests.test_delete_all_usb_partitions;run-3f2b: PASS (1.04 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2023-09-27T22.52-6660df3/results.html
JOB TIME   : 16.06 s

File: avocado/utils/disk.py [hunk 4]
```diff
@@ -311,3 +311,115 @@ def get_disk_mountpoint(device):
...
+        size = get_size_of_disk(disk_name) / 1073741824
+        size = size / num_of_par
+        size = str(size) + "G"
+    partitions = [
+        "size= +" + size if val != 3 else "type=5" for val in range(0, num_of_par + 1)
+    ]
+    disk_partition_file = (
+        "/tmp/creat_partition" + process.run("date '+%d-%m-%y_%T'").stdout_text.strip()
+    )
+    if not os.path.isfile(disk_partition_file):
```
Comment (abdhaleegit) [inline avocado/utils/disk.py:339]: where do you keep the partition file ?   better to  add this as parameter to be passed ..  instead hardcoding path

Comment (abdhaleegit) [reply avocado/utils/disk.py:339]: got it, you create based on number of partitions given

**Review by abdhaleegit — APPROVED** [review]
LGTM

**Comment by abdhaleegit** [comment]
@clebergnu @richtja Please review and merge this

File: avocado/utils/disk.py [hunk 5]
```diff
@@ -311,3 +311,115 @@ def get_disk_mountpoint(device):
             if dev == device:
                 return fs_dir
         return None
+
+
+def create_linux_raw_partition(disk_name, size=None, num_of_par=1):
+    """
+    Creates partitions using sfdisk command
+
+    :param disk_name: disk/device name
```
Comment (PraveenPenguin) [inline avocado/utils/disk.py:320]: can we have only disk name as a doc string as device name can confuse the user and even applies to other places as well

File: avocado/utils/disk.py [hunk 6]
```diff
@@ -311,3 +311,115 @@ def get_disk_mountpoint(device):
...
+        "size= +" + size if val != 3 else "type=5" for val in range(0, num_of_par + 1)
+    ]
+    disk_partition_file = (
+        "/tmp/creat_partition" + process.run("date '+%d-%m-%y_%T'").stdout_text.strip()
+    )
+    if not os.path.isfile(disk_partition_file):
+        process.run("touch " + disk_partition_file)
+    for line in partitions:
+        genio.append_one_line(disk_partition_file, line)
+    part_output = process.getoutput("sfdisk " + disk_name + " < " + disk_partition_file)
```
Comment (PraveenPenguin) [inline avocado/utils/disk.py:343]: what if `sfdisk` disk command not exits, even if the command failed  may have expectation handling

File: avocado/utils/disk.py [hunk 7]
```diff
@@ -311,3 +311,115 @@ def get_disk_mountpoint(device):
...
+
+    :param partition_name: partition absolute path
+    :type partition_name: str
+    """
+    disk_index = re.search(r"\d+", partition_name).start()
+    process.run(
+        "sfdisk --delete "
+        + partition_name[:disk_index]
+        + " "
+        + partition_name[disk_index:]
```
Comment (PraveenPenguin) [inline avocado/utils/disk.py:373]: same here error handling  please

**Review by PraveenPenguin — CHANGES_REQUESTED** [review]

File: avocado/utils/disk.py (hunk 5 above)
Comment (maramsmurthy) [reply avocado/utils/disk.py:320]: ok, As suggested will keep only disk name

File: avocado/utils/disk.py (hunk 6 above)
Comment (maramsmurthy) [reply avocado/utils/disk.py:343]: Sure, Will throw an exception using DiskError Exeption class

File: avocado/utils/disk.py (hunk 7 above)
Comment (maramsmurthy) [reply avocado/utils/disk.py:373]: Sure, Will throw an exception using DiskError Exeption class

**Comment by abdhaleegit** [comment]
@PraveenPenguin please merge if requested comments are implmented

File: avocado/utils/disk.py [hunk 8]
```diff
@@ -13,10 +13,11 @@
 #
 # client/base_utils.py
 #
-# Copyright: 2018 IBM
+# Copyright: 2023 IBM
 # Authors : Praveen K Pandey <praveen@linux.vnet.ibm.com>
 #         : Narasimhan V <sim@linux.vnet.ibm.com>
 #         : Naresh Bannoth <nbannoth@linux.vnet.ibm.com>
+#         : Maram Srimannarayana Murthy <msmurthy@linux.vnet.ibm.com>
```
Comment (PraveenPenguin) [inline avocado/utils/disk.py:20]: ah, I missed it. I did not feel we needed to add an author and copywriter   change

Comment (maramsmurthy) [reply avocado/utils/disk.py:20]: Hi Praveen,
Do you want me to do following
1. Keep copyright as 2018
2. Remove author name in line 20.

Comment (PraveenPenguin) [reply avocado/utils/disk.py:20]: yes correct

**Review by PraveenPenguin — APPROVED** [review]

**Pull request merged**